### DISCUSSION TITLE
LDM Notes for October 7th, 2024

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#8516] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2024/LDM-2024-10-07.md

### Agenda

- Extensions

**Comment by DavidArno** [comment]
> One thing we clarified during discussion is that, even though we've arrived at the idea that extension types aren't actually types and can't be the type of an instance variable today, we haven't ruled out that extension types could implement interfaces in the future, allowing the user to implicitly treat the underlying type as having implemented that interface without having to do any casts.

Hmm, that sounds very much like structural typing to me. Extensions for the win!💯

**Reply by 333fred** [reply]
That isn't structural typing.

**Reply by DavidArno** [reply]
If "_...extension types could implement interfaces in the future, allowing the user to implicitly treat the underlying type as having implemented that interface without having to do any casts_" isn't structural typing, what is it? It looks like a pretty good definition of structural typing to me. 🤷

**Reply by 333fred** [reply]
That's standard type classes. There's no "if it looks like a duck, it's a duck" going on in that sentence; someone will still have had to define the extension type that implements that interface.

**Reply by DavidArno** [reply]
Duck typing is not structural typing. 

But I think I get what you mean. My assumption reading what you wrote was that I could potentially in the future do the following:
```cs
class Foo
{
    void Bar() => etc
}

interface IFoo
{
    void Bar();
}

extension Baz of Foo : IFoo;
```
And because Foo already implements IFoo, Baz doesn't need to declare anything. 

Judging by your reaction though, that isn't what you were saying.

**Reply by FaustVX** [reply]
It will probably work like that as it's already happening with `class`es currently
```C#
class Foo
{
    public void Bar() {}
}

interface IFoo
{
    public void Bar();
}

class Bar : Foo, IFoo;
```
[sharplab.io](https://sharplab.io/#v2:D4AQTABAYg9jCwAoA3ki6IgMyYCwQCEBDAJwAoBKCZAXyTsSQEsA7AFwFMSAzIgYw4QAkrAQo0GbHkKlKAbnpIk4GSQgAuaHAA0w0XKA)

**Reply by HaloFour** [reply]
I think that fits in more with the description of a type class than duck typing.  Either way, I _think_ the behavior that you describe would be possible.  I am assuming that you wouldn't be required to explicitly plumb the extension to call the existing `Bar` method, that the compiler would be able to do that for you.

**Reply by DavidArno** [reply]
But that begs the question, why not? If `E` extends `C` and implements `I` then as long as `E⋅C` implements `I`, it shouldn't matter which of `E` or `C` does the actual implementing.

**Reply by CyrusNajmabadi** [reply]
> And because Foo already implements IFoo, Baz doesn't need to declare anything.

That's def not duck typing.  That's literally just implementing the interface.  Duck typing would be saying that even though the type/extension doesn't implement the interface, it's still a suitable value because the signatures match up and that happens automagically.  Here, there is no magic.  The code is saying: this does implement the interface and then it does :)

**Reply by DavidArno** [reply]
> > And because Foo already implements IFoo, Baz doesn't need to declare anything.
> 
> That's def not duck typing. ..

Duck typing is not structural typing, eg [Duck typing is similar to, but distinct from, structural typing](https://en.wikipedia.org/wiki/Duck_typing#:~:text=Duck%20typing%20is%20similar%20to,that%20is%20accessed%20during%20runtime.)

**Reply by jnm2** [reply]
If I understand structural typing correctly, that would be the ability to pass the class as an interface it doesn't implement, _without_ writing an extension to provide the implementation (either implicitly or explicitly). In other words, by matching the structure of the class directly to the structure of the interface.

Once an extension is in play, the extension is just an extension of the original class. So the extension implementing the interface is equivalent to the original class implementing it. Converting the class to the interface at that point is not structurally typed.

**Reply by 333fred** [reply]
> But that begs the question, why not? If `E` extends `C` and implements `I` then as long as `E⋅C` implements `I`, it shouldn't matter which of `E` or `C` does the actual implementing.

Something needs to provide the vtable that allows virtual dispatch to work. That something cannot be the original type, since it didn't implement `I`, so it must be `E`.

**Reply by CyrusNajmabadi** [reply]
> With duck typing, an object is of a given type if it has all [methods](https://en.wikipedia.org/wiki/Method_(computer_programming)) and properties required by that type

This is not hte case here.  An object will not be of a given type, unless it actually implements that type (either itself, or through an extension).  The presence of all the methods/properties requires will not be enough.

**Reply by DavidArno** [reply]
> Something needs to provide the vtable that allows virtual dispatch to work. That something cannot be the original type, since it didn't implement `I`, so it must be `E`.

But that appears to contradict what you said before, "*allowing the user to implicitly treat the underlying type as having implemented that interface without having to do any casts*". I read that as you saying that the original type can be viewed as now implementing `I`. `E` is just extending `C`, so if `E` adds nothing new, its vtable should be that of `C`, surely?

**Reply by 333fred** [reply]
> But that appears to contradict what you said before, "_allowing the user to implicitly treat the underlying type as having implemented that interface without having to do any casts_". I read that as you saying that the original type can be viewed as now implementing `I`. `E` is just extending `C`, so if `E` adds nothing new, its vtable should be that of `C`, surely?

It doesn't add nothing new. It adds an implementation of `I`, so its VTable cannot be empty. Even if the compiler decided to allow that case to work seamlessly from the user's perspective (appearing to need to write nothing extra), the compiler would need to emit a bridge method to implement that interface. When you convert the underlying type to that interface, some kind of witness will need to be introduced.

**Reply by DavidArno** [reply]
OK, so I could do:
```cs
sealed class Foo
{
    void Bar() => etc
}

interface IFoo
{
    void Bar();
}

extension Baz of Foo : IFoo
{
    void Bar() => Foo.Bar() // syntax not important
}
```
And as you say, the compiler could emit that automatically. If it did emit it automatically, then we now have `Foo` implementing `IFoo` after the event. 

Make the compiler do that and we have that 💯 ...

**Reply by 333fred** [reply]
Sure, if we added automatic emitting of the `extension` itself, we'd have structural typing. But we actively do not want that at the moment, which is why we long ago moved away from shapes.

**Reply by DavidArno** [reply]
> Sure, if we added automatic emitting of the `extension` itself, we'd have structural typing. But we actively do not want that at the moment, which is why we long ago moved away from shapes.

Fair enough. And should people want structural typing via this method, they can always use a source generator to generate those mappings to implement the interface in a generated extension.
